docs/docs/usage/oauth.md

```diff
@@ -0,0 +1,70 @@
+---
+sidebar_position: 5
+---
+
+# OAuth Authentication
+
+This page contains details about using OAuth 2 in Immich.
+
+## Overview
+
+Immich supports 3rd party authentication via [OpenID Connect][oidc] (OIDC), an identity layer built on top of OAuth2. OIDC is supported by most identity providers, including:
+
+- [Authentik](https://goauthentik.io/integrations/sources/oauth/#openid-connect)
+- [Authelia](https://www.authelia.com/configuration/identity-providers/open-id-connect/)
+- [Okta](https://www.okta.com/openid-connect/)
+- [Google](https://developers.google.com/identity/openid-connect/openid-connect)
+
+## Prerequisites
+
+Before enabling OAuth in Immich, a new client application needs to be configured in the 3rd-party authentication server. While the specifics of this setup vary from provider to provider, the general approach should be the same.
+
+1. Create a new (Client) Application
+
+   1. The **Provider** type should be `OpenID Connect` or `OAuth2`
+   2. The **Client type** should be `Confidential`
+   3. The **Application** type should be `Web`
+   4. The **Grant** type should be `Authorization Code`
+
+2. Configure Redirect URIs/Origins
+
+   1. The **Sign-in redirect URIs** should include:
+
+      - All URLs that will be used to access the login page of the Immich web client (eg. `http://localhost:2283/auth/login`, `http://192.168.0.200:2283/auth/login`, `https://immich.example.com/auth/login`)
+
+## Enable OAuth
+
+Once you have a new OAuth client application configured, Immich can be configured using the following environment variables:
+
+| Key                      | Type    | Default              | Description                                                               |
+| ------------------------ | ------- | -------------------- | ------------------------------------------------------------------------- |
+| OAUTH_ENABLED            | boolean | false                | Enable/disable OAuth2                                                     |
+| OAUTH_ISSUER_URL         | URL     | (required)           | Required. Self-discovery URL for client (from previous step)              |
+| OAUTH_CLIENT_ID          | string  | (required)           | Required. Client ID (from previous step)                                  |
+| OAUTH_CLIENT_SECRET      | string  | (required)           | Required. Client Secret (previous step                                    |
+| OAUTH_SCOPE              | string  | openid email profile | Full list of scopes to send with the request (space delimited)            |
+| OAUTH_TOKEN_RESPONSE_ALG | string  | RS256                | Algorithm used to sign the returned ID Token                              |
```

Comment:
Get this from `id_token_signing_alg_values_supported`, require an specific env var if `HS...` because you need a shared secret. Use the key from JWKS if `RS...`

Reply:
Ehm, not quite. Why the focus on `HS256` here? Your current implementation does not work with `HS256` (because it's shared secret, and you set the token secret nowhere). It should be `RS256` (or other RS variants) only. Also, `none` should not be allowed.

Reply:
You can allow for `HS256` if you also include a "token shared secret env var". If it's an RSA variant (like `RS256`) you can use JWKS. Keep in mind that there are many different signing algorithms here: https://www.scottbrady91.com/jose/jwts-which-signing-algorithm-should-i-use

Reply:
OK, I think I understand now. I had not selected a signing key in authentik, so it defaulted to HS256 and used the client secret as the signing key:

https://github.com/goauthentik/authentik/blob/3306003f0e0ac350eaecf3ba3584fcf6c6f4a4d1/authentik/providers/oauth2/models.py#L258-L260

If I pick a key then everything worked as expected. It sounds like you are also saying that I could add support for specifying the secret as an environment variable and that would also make this signing method work.

Reply:
It looks like the `openid-client` library will use the `client_secret` to verify the token if it was signed with `HS265`, but that has to be specified in the client metadata.

Reply:
Sorry for the late comment, but yes exactly right. Afaik using the client_secret is not a standard so you need some way to specify it seperately if necessary.

Did this get fixed or did the PR get merged a bit too enthousiastically? :P

Reply:
As far as I can tell, the library I'm using does not provide a mechanism to pass in a separate secret for validating the `HS` algorithms, so there's no easy way to support this specific feature at the moment.

It's merged since this seems like a small use case that will probably not affect very many users. If it becomes an issue, we can open an issue to track it. The fix would probably involve updating the upstream library (or using a different one), and I don't think it makes sense to _not_ merge this PR while we wait for that to happen.